src/orchestrator/chat_completions_detection.rs

```diff
@@ -1,27 +1,478 @@
+                                                            result,
+                                                        )
+                                                    })
+                                                    .collect::<Vec<_>>(),
+                                            };
+                                            Ok(input_detection_result)
+                                        } else {
+                                            Ok(InputDetectionResult {
+                                                message_index: idx,
+                                                results: vec![],
+                                            })
+                                        }
+                                    }
+                                    Err(error) => Err(error),
+                                }
+                            })
+                        })
+                        .collect::<Vec<_>>()
+                }
+                _ => unimplemented!(),
+            }
+        })
+        .collect::<Vec<_>>();
+
+    let detections = try_join_all(tasks)
+        .await?
+        .into_iter()
+        .collect::<Result<Vec<_>, Error>>()?
+        .into_iter()
+        .filter(|detection| !detection.results.is_empty())
+        .collect::<Vec<InputDetectionResult>>();
+
+    Ok((!detections.is_empty()).then_some(detections))
+}
+
+/// Function to filter messages based on individual detectors
+/// Returns a HashMap of detector id to filtered messages
+fn preprocess_chat_messages(
+    ctx: &Arc<Context>,
+    detectors: &HashMap<String, DetectorParams>,
+    messages: Vec<ChatMessageInternal>,
+) -> Result<HashMap<String, Vec<ChatMessageInternal>>, Error> {
+    detectors
+        .iter()
+        .map(
+            |(detector_id, _)| -> Result<(String, Vec<ChatMessageInternal>), Error> {
+                let ctx = ctx.clone();
+                let detector_id = detector_id.clone();
+                let detector_config = ctx
+                    .config
+                    .detectors
+                    .get(&detector_id)
+                    .unwrap_or_else(|| panic!("detector config not found for {}", detector_id));
+                let detector_type = &detector_config.r#type;
+                // Filter messages based on detector type
+                let value = match detector_type {
+                    DetectorType::TextContents => {
+                        match content::filter_chat_message(messages.clone()) {
+                            Ok(filtered_messages) => Ok(filtered_messages),
+                            Err(e) => Err(Error::Validation(e.to_string())),
+                        }
+                    }
+                    _ => unimplemented!(),
+                }?;
+                Ok((detector_id, value))
+            },
+        )
+        .collect::<Result<HashMap<String, Vec<ChatMessageInternal>>, Error>>()
```

Comment:
nit: type is inferred from return type
```suggestion
        .collect()
```